### PR TITLE
group and ungroup actions, change name

### PR DESCRIPTION
**Pull request by milg15**
Handles groups and ungroups but stills need improvements when it comes to nested groups.

### CODE REVIEW

**Comment by milg15** [comment]
I think there is some code that could be move to util. The sorting array one and the flattening array might be good to move to a different file maybe as a selector(?).

File: packages/app/src/containers/SwatchesGroups.tsx [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+export default memo(function SwatchesGroups() {
+  const [state, dispatch] = useApplicationState();
+
+  const swatches = useShallowArray(useSelector(Selectors.getSharedSwatches));
+  const selectedGroup = state.selectedGroupSwatch;
+
+  const flatSwatchGroup = useMemo(() => {
+    const flat = flatten(createSwatchTree(swatches), '', -1);
```
Comment (dabbott) [inline packages/app/src/containers/SwatchesGroups.tsx:103]: What does the -1 do here?

File: packages/app/src/containers/SwatchesGroups.tsx [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  });
+  return root;
+}
+
+function flatten(
+  swatch: SwatchGroup,
+  parent: string,
+  depth: number,
+): SwatchTitles[] {
+  const full = (parent ? parent + '/' : '') + swatch.name;
```
Comment (dabbott) [inline packages/app/src/containers/SwatchesGroups.tsx:84]: Maybe rename `full` => `path`

File: packages/noya-state/src/reducers/application.ts [hunk 3]
```diff
@@ -1000,9 +1007,11 @@ export function reducer(
             : state.sketch.document.layerStyles?.objects ?? [];
 
         array.forEach((object: Sketch.Swatch | Sketch.SharedStyle) => {
-          if (ids.includes(object.do_objectID)) {
-            object.name = name;
-          }
+          if (!ids.includes(object.do_objectID)) return;
+          const group = object.name.split('/').slice(0, -1).join('/');
```
Comment (dabbott) [inline packages/noya-state/src/reducers/application.ts:1011]: Can this be simplified using the new `dirname` util?

File: packages/noya-state/src/reducers/application.ts [hunk 4]
```diff
@@ -1114,6 +1123,33 @@ export function reducer(
...
+      const [, id, value] = action;
+
+      const ids = typeof id === 'string' ? [id] : id;
+
+      return produce(state, (state) => {
+        const array = state.sketch.document.sharedSwatches?.objects ?? [];
+
+        array.forEach((object: Sketch.Swatch) => {
+          if (!ids.includes(object.do_objectID)) return;
+          const values = object.name.split('/');
```
Comment (dabbott) [inline packages/noya-state/src/reducers/application.ts:1142]: Can this function be simplified with the new `basename` and `dirname` utils?

File: packages/app/src/containers/SwatchesGroups.tsx [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    const aName = a[key].toUpperCase();
+    const bName = b[key].toUpperCase();
+
+    return aName > bName ? 1 : aName < bName ? -1 : 0;
+  });
+
+type SwatchGroup = {
+  name: string;
+  swatches: Sketch.Swatch[];
+  children?: SwatchGroup[];
```
Comment (dabbott) [inline packages/app/src/containers/SwatchesGroups.tsx:34]: Does it help to make children optional?

**Review by dabbott — CHANGES_REQUESTED** [review]
I like the idea of making some utils. Let's make:

- `noya-utils/src/sortBy` (rename `sortArray` => `sortBy`) - Anything in `noya-utils` needs tests! You can run tests just for this package with `yarn test noya-utils`
- `app/src/utils/createSwatchTree` - We will make this generic soon but for now we can just move it. I don't think it's general enough to move to the `noya-utils`
- I'm not sure about flatten yet, but you could move it if you'd like (into app/src/utils/). And maybe rename to `flattenTree`.

I added 2 new utils directly to this branch: `dirname` and `basename`. These mirror the node.js APIs with the same names. We can hopefully use these to simplify (some of) the `.split('/')` and `.join('/')` stuff. You can take a look at the tests to see how they work. I used `basename` in `NameInspector` to test it too.

I also added a pre-commit hook to prevent pushing to GitHub unless TypeScript compiles successfully.

I might have another round of feedback once we make these updates.

**Remember to pull, since I made changes on this branch**

File: packages/app/src/containers/SwatchesGroups.tsx (hunk 5 above)
Comment (milg15) [reply packages/app/src/containers/SwatchesGroups.tsx:34]: I was optional for how a function was working before but now it doesn't need to be👍

File: packages/noya-designsystem/src/components/GridView.tsx [hunk 6]
```diff
@@ -147,12 +170,40 @@ function GridViewRoot({ children, onClick }: GridViewRootProps) {
...
+interface GridViewSectionHeaderProps {
+  title: string;
+}
+
+function GridViewSectionHeader({ title }: GridViewSectionHeaderProps) {
+  const groupedTitle = title.split('/');
+
+  return (
+    <SectionHeaderContainer>
+      {groupedTitle.map((title, index) => {
```
[the remark this answers is not in the document]
Comment (milg15) [reply packages/noya-designsystem/src/components/GridView.tsx:191]: Oh! Awesome! I added the function but had to run it with this path

`import withSeparatorElements from '../../../app/src/utils/withSeparatorElements';`
😅 

I don't think it's optimal, it's there any better way of reaching this file?

File: packages/app/src/components/theme/SwatchesGrid.tsx [hunk 7]
```diff
@@ -1,55 +1,131 @@
...
-  const sortedSwatches = useMemo(
-    () =>
-      [...swatches].sort((a, b) => {
-        const aName = a.name.toUpperCase();
-        const bName = b.name.toUpperCase();
-
-        return aName > bName ? 1 : aName < bName ? -1 : 0;
-      }),
-    [swatches],
+  const swatchesGrouped = useMemo(() => {
```
[the remark this answers is not in the document]
Comment (milg15) [reply packages/app/src/components/theme/SwatchesGrid.tsx:34]: Yes! I'll join with the new swatch group file

File: packages/noya-state/src/reducers/application.ts (hunk 3 above)
Comment (milg15) [reply packages/noya-state/src/reducers/application.ts:1011]: yes !

**Comment by milg15** [comment]
I still need to add more test for the sortBy. I think it's better now that the createSwatchTree it's now separate. 

I kept the flatten functions in the files because they return differen results in my opinion.

File: packages/noya-designsystem/src/components/GridView.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (dabbott) [reply packages/noya-designsystem/src/components/GridView.tsx]: Good catch! I forgot that was in another package. Let's move withSeparatorElements.ts => `noya-designsystem/src/utils/withSeparatorElements.ts` and export it from `noya-designsystem/src/index.tsx`

File: packages/app/src/components/inspector/NameInspector.tsx [hunk 8]
```diff
@@ -8,10 +9,11 @@ interface Props {
 }
 
 export default memo(function NameInspector({ names, onNameChange }: Props) {
-  const firstName = names[0];
+  const firstName = delimitedPath.basename(names[0]);
```
Comment (dabbott) [inline packages/app/src/components/inspector/NameInspector.tsx:12]: I just realized we're handling `names` and `onNameChange` inconsistently:

- `names` are all full paths, like `["group/nestedGroup/a"]`
- `onNameChange` only gets called with the base name, `"a"`.

I think we should use either full path or base name for both props. `NameInspector` doesn't need to know the full path, cause it doesn't show it, so I would transform `names` into an array of base name, `["a"]`, before passing it in as a prop. It might mean a little duplicated code in the other inspectors, but we can improve that in other ways

File: packages/app/src/components/theme/SwatchesGrid.tsx [hunk 9]
```diff
@@ -1,55 +1,136 @@
...
+      }
+    },
+    [onSelectSwatch, onGroupSwatch, onDeleteSwatch, selectedSwatchIds],
+  );
 
-        return aName > bName ? 1 : aName < bName ? -1 : 0;
-      }),
+  const handleOnContextMenu = useCallback(
+    (id: string) => {
+      if (selectedSwatchIds.length && selectedSwatchIds.includes(id)) return;
```
Comment (dabbott) [inline packages/app/src/components/theme/SwatchesGrid.tsx:79]: Do we need to check `selectedSwatchIds.length` here?

File: packages/app/src/components/theme/SwatchesGrid.tsx [hunk 10]
```diff
@@ -1,55 +1,136 @@
...
+function flatten(
+  swatch: SwatchGroup,
+  parent: string,
+): {
+  path: string;
+  swatches: Sketch.Swatch[];
+}[] {
+  const path = (parent ? parent + '/' : '') + swatch.name;
+
+  return Array.prototype.concat.apply(
```
Comment (dabbott) [inline packages/app/src/components/theme/SwatchesGrid.tsx:20]: Can we do

```ts
return [
  {
    path: path,
    swatches: swatch.swatches,
  },
  ...swatch.children.map((x) => flatten(x, path)),
);
```

File: packages/app/src/components/theme/SwatchesGrid.tsx [hunk 11]
```diff
@@ -1,55 +1,136 @@
...
+import { ContextMenu, GridView, sketchColorToRgba } from 'noya-designsystem';
 import { SelectionType } from 'noya-state';
-import { memo, useMemo } from 'react';
+import { memo, useMemo, useCallback, Fragment } from 'react';
+import { SwatchGroup, createSwatchTree } from '../../utils/createSwatchTree';
+import { sortBy } from 'noya-utils';
 import ColorSwatch from './ColorSwatch';
 
+function flatten(
+  swatch: SwatchGroup,
```
Comment (dabbott) [inline packages/app/src/components/theme/SwatchesGrid.tsx:12]: Rename `swatch` => `swatchGroup`

File: packages/app/src/components/theme/SwatchesGrid.tsx [hunk 12]
```diff
@@ -1,55 +1,136 @@
...
-            <ColorSwatch value={item.value} />
-          </GridView.Item>
+          <Fragment key={index}>
+            {group.path && <GridView.SectionHeader title={group.path} />}
+            <GridView.Section>
+              {sortedSwatches.map((item) => {
+                const color = sketchColorToRgba(item.value);
+                const hex = rgbaToHex(color);
+                const alphaPercent = `${Math.round(color.a * 100)}%`;
+                const name = item.name.split('/').pop() || '';
```
Comment (dabbott) [inline packages/app/src/components/theme/SwatchesGrid.tsx:109]: Can use `basename` here I think

File: packages/app/src/components/theme/SwatchesGrid.tsx [hunk 13]
```diff
@@ -1,55 +1,136 @@
...
   );
 
   return (
     <GridView.Root onClick={() => onSelectSwatch(undefined, 'replace')}>
-      {sortedSwatches.map((item) => {
-        const color = sketchColorToRgba(item.value);
-        const hex = rgbaToHex(color);
-        const alphaPercent = `${Math.round(color.a * 100)}%`;
+      {flatSwatchGroup.map((group, index) => {
+        const sortedSwatches = sortBy(group.swatches, 'name');
```
Comment (dabbott) [inline packages/app/src/components/theme/SwatchesGrid.tsx:99]: I think these should already be sorted, cause we sorted by 'path' above already

File: packages/app/src/containers/SwatchesGroups.tsx [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+}));
+
+type SwatchTitles = {
+  name: string;
+  path: string;
+  depth: number;
+};
+
+function flatten(
+  swatch: SwatchGroup,
```
Comment (dabbott) [inline packages/app/src/containers/SwatchesGroups.tsx:31]: Rename `swatch` => `SwatchGroup`

File: packages/app/src/containers/SwatchesGroups.tsx [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+  parent: string,
+  depth: number,
+): SwatchTitles[] {
+  const path = (parent ? parent + '/' : '') + swatch.name;
+
+  return Array.prototype.concat.apply(
```
Comment (dabbott) [inline packages/app/src/containers/SwatchesGroups.tsx:37]: Same as comment above, we can use array spread instead of `concat.apply`

File: packages/noya-state/src/reducers/application.ts [hunk 16]
```diff
@@ -1148,6 +1186,7 @@ export function createInitialState(sketch: SketchFile): ApplicationState {
     selectedObjects: [],
     selectedSwatchIds: [],
     selectedLayerStyleIds: [],
+    selectedGroupSwatch: '',
```
Comment (dabbott) [inline packages/noya-state/src/reducers/application.ts:1189]: Rename `selectedGroupSwatch` => `selectedSwatchGroup`

File: packages/app/src/utils/createSwatchTree.ts [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+    name: '',
+    swatches: [],
+    children: [],
+  };
+
+  function getGroup(pathComponents: string[]): SwatchGroup {
+    let group = root;
+    while (pathComponents.length > 0) {
+      const component = pathComponents.shift()!;
+      if (!group.children) continue;
```
Comment (dabbott) [inline packages/app/src/utils/createSwatchTree.ts:20]: Do we need to check `!group.children`?

File: packages/noya-state/src/reducers/application.ts [hunk 18]
```diff
@@ -1114,6 +1123,35 @@ export function reducer(
...
+      return produce(state, (state) => {
+        const array = state.sketch.document.sharedSwatches?.objects ?? [];
+
+        array.forEach((object: Sketch.Swatch) => {
+          if (!ids.includes(object.do_objectID)) return;
+
+          const prevGroup = dirname(object.name);
+          const name = basename(object.name);
+
+          const group = value
```
Comment (dabbott) [inline packages/noya-state/src/reducers/application.ts:1147]: It looks like this code should join the different path segments only if they exist, but I find it a little hard to follow. Will this work?

```ts
const newName = [prevGroup, value, name].filter(x => !!x).join('/')
object.name = newName
```

Possibly a future util, `delimitedPath.join`

**Review by dabbott — CHANGES_REQUESTED** [review]
This is looking good! Mostly some small stuff now.

File: packages/app/src/components/theme/SwatchesGrid.tsx (hunk 10 above)
Comment (milg15) [reply packages/app/src/components/theme/SwatchesGrid.tsx:20]: It's not working. I need to use the .apply, maybe there is another way it gives me a typescript error

Comment (milg15) [reply packages/app/src/components/theme/SwatchesGrid.tsx:20]: It gives me a typescript error 🤔

File: packages/app/src/components/inspector/NameInspector.tsx (hunk 8 above)
Comment (milg15) [reply packages/app/src/components/inspector/NameInspector.tsx:12]: Oh! It took me a second to understand but I got it, the action uses ids to change the name so i don't need the full name

**Comment by milg15** [comment]
Other than the ts error I mentioned I added all the changes.

File: packages/app/src/components/theme/SwatchesGrid.tsx (hunk 10 above)
Comment (dabbott) [reply packages/app/src/components/theme/SwatchesGrid.tsx:20]: What if we do `flatMap` instead of `map`?

**Comment by dabbott** [comment]
👍  I think I got the types wrong, but if we do a `flatMap` it should work

File: packages/app/src/components/theme/SwatchesGrid.tsx [hunk 19]
```diff
@@ -1,57 +1,136 @@
...
 import { SelectionType } from 'noya-state';
-import { memo, useMemo } from 'react';
+import { memo, useMemo, useCallback, Fragment } from 'react';
+import { SwatchGroup, createSwatchTree } from '../../utils/createSwatchTree';
+import { sortBy } from 'noya-utils';
 import ColorSwatch from './ColorSwatch';
+import { delimitedPath } from 'noya-utils';
+
+function flatten(
+  SwatchGroup: SwatchGroup,
```
Comment (dabbott) [inline packages/app/src/components/theme/SwatchesGrid.tsx:13]: Argument names should usually be lowercase, `swatchGroup`, but I'll change it another time

File: packages/noya-utils/src/__tests__/sortBy.test.ts [hunk 20]
```diff
@@ -0,0 +1,5 @@
+import { sortBy } from '../index';
+
+test('empty array', () => {
+  expect(sortBy([], '')).toEqual([]);
```
Comment (dabbott) [inline packages/noya-utils/src/__tests__/sortBy.test.ts:4]: I'm going to merge this, but I'll create a follow-up issue to add some more tests for `sortBy`

File: packages/noya-utils/src/sortBy.ts [hunk 21]
```diff
@@ -0,0 +1,7 @@
+export const sortBy = (array: any[], key: string) =>
```
Comment (dabbott) [inline packages/noya-utils/src/sortBy.ts:1]: I'm also going to try to improve the `any` type here

**Review by dabbott — APPROVED** [review]

File: packages/app/src/components/theme/SwatchesGrid.tsx (hunk 19 above)
Comment (milg15) [reply packages/app/src/components/theme/SwatchesGrid.tsx:13]: oh I'll keep it in mid !
